### PR TITLE
CZ inherits from Controlled

### PR DESCRIPTION
**Pull request by BorjaRequena**
**Context:**
Controlled qubit gates, such as `qml.CZ`, are manually implemented and scattered across files (e.g., in `pennylane/ops/qubit/non_parametric_ops.py`). We now have the classes `Controlled` and `ControlledOp`, which handle the logic for controlled operations. Thus, it would be best to harmonize all controlled operations, putting them together and inheriting from `ControlledOp`, which will significantly reduce the amount of redundant code.

**Description of the Change:**
I am reimplementing the controlled Pauli Z operation `qml.CZ` to be a subclass of `ControlledOp`. Additionally, I am moving it from its original location, `pennylane/ops/qubit/non_parametric_ops.py`, to `pennylane/ops/op_math/controlled_ops.py`, where all the controlled operations will eventually lie.

**Benefits:**
This results in less duplicate code and increased functionalities inherited from `Controlled` for `CZ`. For instance, it now has a sparse matrix representation, a `decomposition`, and a `pow` method that can handle arbitrary exponents (before it was just integers).

**Possible Drawbacks:**
While we migrate all the controlled operations, the code may be a bit messy.

**Related GitHub Issues:**
#1447

### CODE REVIEW

**Comment by BorjaRequena** [comment]
Some additional comments and questions:

**Changes to `Controlled`**
While the `Controlled` class helps to design controlled operations, I have had to reimplement a few methods, such as `adjoint` and `pow`. This is because the `Controlled` class uses `self.__class__` in these methods, making it such that whenever we create a subclass out of it they will no longer work. For instance, to create the adjoint operator, we always want to create a `Controlled(self.base.adjoint(), ...)`, not a `CZ(self.base.adjoint(), ...)`. I believe that replacing `self.__class__` there for `Controlled` would help reduce the amount of redundant code.

**__repr__**
With the new `qml.CZ` class, the `__repr__` is `Controlled(PauliZ(wires), control_wires)`. Do we want to change it to be `CZ` and match the previous behaviour?

**Operator representations**
We now have `has_diagonalizing_gates = True` for `CZ`, but `diagonalizing_gates` returns an empty list. Is this the intended behaviour? Should it be `has_diagonalizing_gates = False`?

**Tests**
I have also moved the tests belonging to the `qml.CZ` class to `test_controlled_ops`, preparing the test classes to take additional controlled operations, as in the original tests. Additionally, I have created tests for the new `decomposition` in terms of a controlled `PhaseShift`, and for arbitrary exponents in `pow`. However, testing `pow` for integer exponents, I had to replace the comparison between classes (assessing equality between `op.__class__`) to check that the `op.matrix()` is the same. This is because `qml.CZ`'s class is `CZ`, but after `pow` it becomes `Controlled(PauliZ)`, which is the same in practice but not semantically. Any thoughts on how to improve this?

**Comment by albi3ro** [comment]
[sc-38096]

File: tests/ops/op_math/test_controlled_ops.py [hunk 1]
```diff
@@ -428,3 +441,140 @@ def test_unitary_check(self):
...
             )
+
+
+class TestOperations:
+    @pytest.mark.parametrize("op_cls, _", NON_PARAMETRIZED_OPERATIONS)
+    def test_nonparametrized_op_copy(self, op_cls, _, tol):
+        """Tests that copied nonparametrized ops function as expected"""
+        op = op_cls(wires=0 if op_cls.num_wires is AnyWires else range(op_cls.num_wires))
+        copied_op = copy.copy(op)
+        np.testing.assert_allclose(op.matrix(), copied_op.matrix(), atol=tol)
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:452]: Looking at `test_non_parametric_ops.py` now, I see this is how we were indeed testing the `copy` method.  Probably not the best way to do it.  

Since this is just moving pre-existing code, I won't complain too much, but I would prefer if we actually tested the attributes that needed to be copied, like `data`, `_wires`, and `_hyperparameters`.

File: tests/ops/op_math/test_controlled_ops.py [hunk 2]
```diff
@@ -428,3 +441,140 @@ def test_unitary_check(self):
...
+class TestDecompositions:  # pylint: disable=too-few-public-methods
+    def test_CZ_decomposition(self, tol):
+        """Tests that the decomposition of the CZ gate is correct"""
+        op = qml.CZ(wires=[0, 1])
+        res = op.decomposition()
+
+        assert len(res) == 1
+        assert isinstance(res[0], qml.ops.ControlledOp)
+        assert res[0].base.name == "PhaseShift"
+        assert res[0].data[0] == np.pi
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:473]:
```suggestion
        assert qml.equal(qml.ctrl(qml.PhaseShift(np.pi/2, wires=1), 0), res[0])
```

We can do this now.

File: tests/ops/op_math/test_controlled_ops.py [hunk 3]
```diff
@@ -428,3 +441,140 @@ def test_unitary_check(self):
...
+
+    def test_CZ(self):
+        """Test the PauliZ _controlled method."""
+        out = qml.CZ(wires=[0, 1])._controlled("a")  # pylint: disable=protected-access
+        assert qml.equal(out, qml.CCZ(("a", 0, 1)))
+
+
+class TestSparseMatrix:  # pylint: disable=too-few-public-methods
+    @pytest.mark.parametrize("op, mat", SPARSE_MATRIX_SUPPORTED_OPERATIONS)
+    def test_sparse_matrix(self, op, mat):
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:536]:
```suggestion
    def test_sparse_matrix(self, op, mat):
        """Tests the sparse matrix method for operations which support it."""
```

File: tests/ops/op_math/test_controlled_ops.py [hunk 4]
```diff
@@ -428,3 +441,140 @@ def test_unitary_check(self):
...
+        assert all(sparse_mat.indices == expected_sparse_mat.indices)
+
+
+label_data = [
+    (qml.CZ(wires=(0, 1)), "Z"),
+]
+
+
+@pytest.mark.parametrize("op, label", label_data)
+def test_label_method(op, label):
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:552]:
```suggestion
def test_label_method(op, label):
    """Tests that the label method gives the expected r esult."""
```

File: tests/ops/op_math/test_controlled_ops.py [hunk 5]
```diff
@@ -428,3 +441,140 @@ def test_unitary_check(self):
...
+]
+
+
+@pytest.mark.parametrize("op", involution_ops)
+def test_adjoint_method(op):
+    adj_op = copy.copy(op)
+    for _ in range(4):
+        adj_op = adj_op.adjoint()
+
+        assert adj_op.name == op.name
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:580]:
```suggestion
        assert qml.equal(adj_op, op)
```
You would want this test to fail if `op` was something like `CRX`.  The name would still be the same, but the parameter would be different.  This test wouldn't catch that.

File: tests/ops/op_math/test_controlled_ops.py [hunk 6]
```diff
@@ -428,3 +441,140 @@ def test_unitary_check(self):
...
+    assert op.control_wires == control_wires
+
+
+involution_ops = [  # ops who are their own inverses
+    qml.CZ((0, 1)),
+]
+
+
+@pytest.mark.parametrize("op", involution_ops)
+def test_adjoint_method(op):
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_ops.py:575]:
```suggestion
def test_adjoint_method(op):
    """Tests the adjoint method for operations that are their own adjoint."""
```

**Review by albi3ro — COMMENTED** [review]
Mostly just some minor comments on the tests, but other than that I think this looks pretty good. Will probably approve once the minor comments have been addressed.


`CZ` should also be added to this list [here](https://github.com/PennyLaneAI/pennylane/blob/00a1972e4d2d28a8ef3aa657cec718721c7c51ab/pennylane/ops/op_math/__init__.py#L60) which has the list of controlled operators present in the `op_math` module.

File: tests/ops/op_math/test_controlled_ops.py (hunk 1 above)
Comment (BorjaRequena) [reply tests/ops/op_math/test_controlled_ops.py:452]: Maybe we could use `qml.equal` here too? If I understand it correctly, it'll check that these attributes are the same between two operators.

**Review by albi3ro — APPROVED** [review]
Looks good 🎉 Here's an approval from me :)

File: pennylane/devices/default_qubit.py [hunk 7]
```diff
@@ -311,7 +311,7 @@ def _apply_operation(self, state, operation):
         if operation.__class__.__name__ in self._apply_ops:
             shift = int(self._ndim(state) > self.num_wires)
             axes = [ax + shift for ax in self.wires.indices(wires)]
-            return self._apply_ops[operation.base_name](state, axes)
+            return self._apply_ops[operation.__class__.__name__](state, axes)
```
Comment (timmysilv) [inline pennylane/devices/default_qubit.py:314]: are the changes here and in null_qubit some kind of merge with #4130 ? the changes are not the same so I'm unsure

File: pennylane/ops/op_math/controlled.py [hunk 8]
```diff
@@ -368,7 +368,7 @@ def map_wires(self, wire_map: dict):
         new_control_wires = Wires([wire_map.get(wire, wire) for wire in self.control_wires])
         new_work_wires = Wires([wire_map.get(wire, wire) for wire in self.work_wires])
 
-        return self.__class__(
+        return Controlled(
```
Comment (timmysilv) [inline pennylane/ops/op_math/controlled.py:371]: why did you need to make this change?

File: pennylane/ops/qubit/non_parametric_ops.py [hunk 9]
```diff
@@ -1043,103 +1043,6 @@ def is_hermitian(self):
...
-        >>> print(qml.CZ.compute_matrix())
-        [[ 1  0  0  0]
-         [ 0  1  0  0]
-         [ 0  0  1  0]
-         [ 0  0  0 -1]]
-        """
-        return np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, -1]])
-
-    @staticmethod
-    def compute_eigvals():  # pylint: disable=arguments-differ
```
Comment (timmysilv) [inline pennylane/ops/qubit/non_parametric_ops.py:1100]: did you drop this because we want to use Controlled's default behaviour?

File: pennylane/devices/default_qubit.py (hunk 7 above)
Comment (BorjaRequena) [reply pennylane/devices/default_qubit.py:314]: Oh I didn't realize there was another PR changing these lines. I believe the changes in #4130 would also make this work, as we're addressing the same issue. The main problem here was checking that `operation.__class__.__name__` was in the keys and then indexing with `operation.base_name`, so I changed it to be consistent with the `if` statement.

File: pennylane/ops/op_math/controlled.py (hunk 8 above)
Comment (BorjaRequena) [reply pennylane/ops/op_math/controlled.py:371]: These calls to `self.__class__` create issues whenever we change the call signature of the child classes. For instance, in `CZ` we no longer need to provide a `base` argument, which results in an error. I changed this to `Controlled` because this way we have a consistent behaviour regardless of the call signatures.

File: pennylane/ops/qubit/non_parametric_ops.py (hunk 9 above)
Comment (BorjaRequena) [reply pennylane/ops/qubit/non_parametric_ops.py:1100]: I think so. Dropping `compute_eigvals` was suggested by @albi3ro so maybe she can give you proper arguments behind this choice :)

File: pennylane/ops/op_math/controlled.py (hunk 8 above)
Comment (timmysilv) [reply pennylane/ops/op_math/controlled.py:371]: this could cause other operations to have problems, so I'd like to avoid this if possible. For example, if we later define `class ControlledWithCustomization(Controlled)`, then call `map_wires` on an instance of that, it will use the wrong constructor. I'm wondering why this was needed here but not in the `CY` PR - missing tests there, or maybe something here?

Comment (timmysilv) [reply pennylane/ops/op_math/controlled.py:371]: Chatted offline about this. I opened #4139 to update these class methods to use `qml.ctrl` instead. Once that's in, we'll update this PR to just have custom logic for CZ in `ctrl`

Comment (eddddddy) [reply pennylane/ops/op_math/controlled.py:371]: What's the issue with overriding these 4 methods in the subclass?

Comment (timmysilv) [reply pennylane/ops/op_math/controlled.py:371]: it's fine to do, but we'll have to do it for CY as well. And maybe other controlled ops will want custom code in some other way, so the linked PR will make that easier. Plus, we don't want to re-write code that already works just because the interface is a little too unique/snowflakey for the new class

Comment (eddddddy) [reply pennylane/ops/op_math/controlled.py:371]: I agree that it's the same issue for both `CY` and `CZ` (and other non-parametric ops), but that doesn't mean we have to override them more than once.

 I was thinking we could have a subclass of `Controlled`, say `ControlledNonParametric`, that would be the base class for `CY` and `CZ` and overrides these 4 methods. Then nothing needs to be changed in the `Controlled` class.

File: pennylane/ops/op_math/controlled.py [hunk 10]
```diff
@@ -85,9 +85,14 @@ def circuit(x):
...
 
     """
+    custom_controlled_ops = {
+        qml.PauliZ: qml.CZ,
+    }
     control_values = [control_values] if isinstance(control_values, int) else control_values
     control = qml.wires.Wires(control)
 
+    if isinstance(op, tuple(custom_controlled_ops)) and len(control) == 1:
+        return custom_controlled_ops[type(op)](control + op.wires)
```
[the remark this answers is not in the document]
Comment (BorjaRequena) [reply pennylane/ops/op_math/controlled.py:95]: Oh, exactly what I thought!! Thanks <3

**Review by timmysilv — APPROVED** [review]

File: pennylane/ops/op_math/controlled.py [hunk 11]
```diff
@@ -85,9 +86,19 @@ def circuit(x):
...
+    custom_controlled_ops = {
+        qml.PauliZ: qml.CZ,
+    }
+    control_values = [control_values] if isinstance(control_values, (int, bool)) else control_values
     control = qml.wires.Wires(control)
 
+    if (
+        isinstance(op, tuple(custom_controlled_ops))
+        and len(control) == 1
+        and (control_values is None or control_values[0])
```
[the remark this answers is not in the document]
Comment (BorjaRequena) [reply pennylane/ops/op_math/controlled.py:97]: I am not sure about how work wires actually work. I can imagine that, even if we consider work wires, the decomposition of these gates will be in the same wires they are defined, as these extra wires would be optional (?) I wouldn't worry too much about them neither :)
